src/plugins/formulas/formulas.js

```diff
@@ -1,425 +1,197 @@
-   * @returns {*}
-   */
-  onBeforeValueRender(value) {
-    let renderValue = value;
-
-    if (isFormulaExpressionEscaped(renderValue)) {
-      renderValue = unescapeFormulaExpression(renderValue);
-    }
-
-    return renderValue;
-  }
-
-  /**
-   * On before validate listener.
-   *
-   * @private
-   * @param {*} value Value to validate.
-   * @param {Number} row Row index.
-   * @param {Number} prop Column property.
-   */
-  onBeforeValidate(value, row, prop) {
-    const column = this.hot.propToCol(prop);
-    let validateValue = value;
-
-    if (this.hasComputedCellValue(row, column)) {
-      validateValue = this.getCellValue(row, column);
-    }
-
-    return validateValue;
-  }
-
-  /**
-   * `afterSetDataAtCell` listener.
-   *
-   * @private
-   * @param {Array} changes Array of changes.
-   * @param {String} [source] Source of changes.
-   */
-  onAfterSetDataAtCell(changes, source) {
-    if (source === 'loadData') {
-      return;
-    }
-
-    this.dataProvider.clearChanges();
-
-    arrayEach(changes, ([row, column, oldValue, newValue]) => {
-      const physicalColumn = this.hot.propToCol(column);
-      const physicalRow = this.hot.toPhysicalRow(row);
-      let value = newValue;
-
-      if (isFormulaExpression(value)) {
-        value = toUpperCaseFormula(value);
-      }
-
-      this.dataProvider.collectChanges(physicalRow, physicalColumn, value);
-
-      if (oldValue !== value) {
-        this.sheet.applyChanges(physicalRow, physicalColumn, value);
-      }
-    });
-    this.recalculate();
+    // TODO: check if these hooks from the previous version of the plugin are still relevant
+    // this.addHook('afterSetDataAtCell', (...args) => this.onAfterSetDataAtCell(...args));
+    // this.addHook('afterSetDataAtRowProp', (...args) => this.onAfterSetDataAtCell(...args));
+    // this.addHook('beforeColumnSort', (...args) => this.onBeforeColumnSort(...args));
+    // this.addHook('beforeCreateCol', (...args) => this.onBeforeCreateCol(...args));
+    // this.addHook('beforeCreateRow', (...args) => this.onBeforeCreateRow(...args));
+    // this.addHook('beforeRemoveCol', (...args) => this.onBeforeRemoveCol(...args));
+    // this.addHook('beforeRemoveRow', (...args) => this.onBeforeRemoveRow(...args));
+    // this.addHook('beforeValidate', (...args) => this.onBeforeValidate(...args));
+    // this.addHook('beforeValueRender', (...args) => this.onBeforeValueRender(...args));
   }
 
   /**
-   * On before create row listener.
+   * On after remove row listener. TODO: check docs
    *
    * @private
    * @param {Number} row Row index.
    * @param {Number} amount An amount of removed rows.
-   * @param {String} source Source of method call.
    */
-  onBeforeCreateRow(row, amount, source) {
-    if (source === 'UndoRedo.undo') {
-      this.undoRedoSnapshot.restore();
+  onAfterRemoveRow(row, amount) {
+    if (this.hyperformula.isItPossibleToRemoveRows(this.sheetId, [row, amount])) {
```

Comment:
I know that you know. But let's create a written task from this one. All `after*` hooks should have `before*` version with `isItPossibleTo*` method in them.

```ts
onBeforeRemoveRow(row, amount) {
   return this.hyperformula.isItPossibleToRemoveRows(this.sheetId, [row, amount]);
}
```

The reason for this request is that if we do the check in `afterRemoveRow` then the hooks added by our users will be triggered when they should be blocked (action is not possible with formula plugin enabled).

We should add tests for all actions that can be blocked by the HyperFormula engine, AND for a use case when it's a user hook that blocks the `beforeRemoveRow` hook. In that case, the HyperFormula action should not be called. Otherwise, we will have a desynchronization between Handsontable and HyperFormula.

Also, please add this reasoning to PR description for future release notes and documentation purposes. @jansiegel and @scarletfog will appreciate your effort 😄